File: todo.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Task struct {
	ID          int    `json:"id"`
	Title       string `json:"title"`
	Description string `json:"description"`
	Priority    string `json:"priority"` // high, medium, low
	Due         string `json:"due"`
	Completed   bool   `json:"completed"`
}

var tasks []Task
var nextID = 1
const dataFile = "tasks.json"

func load() {
	file, err := os.ReadFile(dataFile)
	if err != nil { return }
	json.Unmarshal(file, &tasks)
	for _, t := range tasks { if t.ID >= nextID { nextID = t.ID + 1 } }
}

func save() {
	data, _ := json.MarshalIndent(tasks, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func addTask() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Название: ")
	title, _ := reader.ReadString('\n')
	title = strings.TrimSpace(title)
	if title == "" { fmt.Println("Ошибка: название нужно"); return }
	fmt.Print("Описание: ")
	desc, _ := reader.ReadString('\n')
	fmt.Print("Приоритет (high/medium/low) [medium]: ")
	prio, _ := reader.ReadString('\n')
	prio = strings.TrimSpace(strings.ToLower(prio))
	if prio != "high" && prio != "low" { prio = "medium" }
	fmt.Print("Дедлайн (ГГГГ-ММ-ДД, необязательно): ")
	due, _ := reader.ReadString('\n')
	due = strings.TrimSpace(due)
	tasks = append(tasks, Task{ID: nextID, Title: title, Description: desc, Priority: prio, Due: due, Completed: false})
	nextID++
	save()
	fmt.Println("✅ Задача добавлена")
}

func listTasks() {
	if len(tasks) == 0 { fmt.Println("Нет задач."); return }
	fmt.Println("\n=== Список задач ===")
	for _, t := range tasks {
		status := " "
		if t.Completed { status = "✓" }
		prioSym := map[string]string{"high":"🔴","medium":"🟡","low":"🟢"}[t.Priority]
		fmt.Printf("[%s] %d: %s %s | %s | %s\n", status, t.ID, t.Title, prioSym, t.Description, t.Due)
	}
}

func toggleComplete() {
	fmt.Print("ID задачи: ")
	var id int
	fmt.Scanln(&id)
	for i, t := range tasks {
		if t.ID == id {
			tasks[i].Completed = !tasks[i].Completed
			save()
			fmt.Println("✓ Статус изменён")
			return
		}
	}
	fmt.Println("Задача не найдена")
}

func deleteTask() {
	fmt.Print("ID задачи для удаления: ")
	var id int
	fmt.Scanln(&id)
	for i, t := range tasks {
		if t.ID == id {
			tasks = append(tasks[:i], tasks[i+1:]...)
			save()
			fmt.Println("✓ Удалено")
			return
		}
	}
	fmt.Println("Не найдено")
}

func filterTasks() {
	fmt.Println("1. Все\n2. Активные\n3. Выполненные\n4. По приоритету")
	fmt.Print("Выбор: ")
	var choice int
	fmt.Scanln(&choice)
	switch choice {
	case 2:
		for _, t := range tasks { if !t.Completed { printTask(t) } }
	case 3:
		for _, t := range tasks { if t.Completed { printTask(t) } }
	case 4:
		fmt.Print("Приоритет (high/medium/low): ")
		var prio string
		fmt.Scanln(&prio)
		for _, t := range tasks { if t.Priority == prio { printTask(t) } }
	default:
		listTasks()
	}
}

func printTask(t Task) {
	fmt.Printf("%d: %s [%s] %s\n", t.ID, t.Title, t.Priority, t.Description)
}

func main() {
	load()
	reader := bufio.NewReader(os.Stdin)
	for {
		fmt.Println("\n=== Todo Manager ===")
		fmt.Println("1. Добавить\n2. Список всех\n3. Выполнить/вернуть\n4. Удалить\n5. Фильтр\n0. Выход")
		fmt.Print("> ")
		opt, _ := reader.ReadString('\n')
		opt = strings.TrimSpace(opt)
		switch opt {
		case "1": addTask()
		case "2": listTasks()
		case "3": toggleComplete()
		case "4": deleteTask()
		case "5": filterTasks()
		case "0": fmt.Println("До свидания!"); return
		default: fmt.Println("Неверный ввод")
		}
	}
}

File: todo.java
import java.io.*;
import java.nio.file.*;
import java.time.LocalDate;
import java.util.*;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.core.type.TypeReference;

class Task {
    public int id;
    public String title;
    public String description;
    public String priority;
    public String due;
    public boolean completed;

    public Task() {}
    public Task(int id, String title, String desc, String priority, String due, boolean completed) {
        this.id = id; this.title = title; this.description = desc; this.priority = priority; this.due = due; this.completed = completed;
    }
}

public class TodoApp {
    private static List<Task> tasks = new ArrayList<>();
    private static int nextId = 1;
    private static final String DATA_FILE = "tasks.json";
    private static ObjectMapper mapper = new ObjectMapper();

    public static void main(String[] args) throws IOException {
        load();
        Scanner sc = new Scanner(System.in);
        while (true) {
            System.out.println("\n=== Java Todo ===");
            System.out.println("1. Добавить\n2. Список всех\n3. Выполнить/вернуть\n4. Удалить\n5. Фильтр (активные/выполненные)\n0. Выход");
            System.out.print("> ");
            String opt = sc.nextLine();
            switch (opt) {
                case "1": addTask(sc); break;
                case "2": listAll(); break;
                case "3": toggleComplete(sc); break;
                case "4": deleteTask(sc); break;
                case "5": filterMenu(sc); break;
                case "0": System.out.println("До свидания!"); return;
                default: System.out.println("Неверно");
            }
        }
    }

    static void load() throws IOException {
        File f = new File(DATA_FILE);
        if (f.exists()) {
            String content = new String(Files.readAllBytes(Paths.get(DATA_FILE)));
            tasks = mapper.readValue(content, new TypeReference<List<Task>>(){});
            nextId = tasks.stream().mapToInt(t -> t.id).max().orElse(0) + 1;
        }
    }

    static void save() throws IOException {
        mapper.writeValue(new File(DATA_FILE), tasks);
    }

    static void addTask(Scanner sc) throws IOException {
        System.out.print("Название: ");
        String title = sc.nextLine().trim();
        if (title.isEmpty()) { System.out.println("Ошибка"); return; }
        System.out.print("Описание: ");
        String desc = sc.nextLine();
        System.out.print("Приоритет (high/medium/low) [medium]: ");
        String prio = sc.nextLine().toLowerCase();
        if (!prio.equals("high") && !prio.equals("low")) prio = "medium";
        System.out.print("Дедлайн (ГГГГ-ММ-ДД, Enter пропустить): ");
        String due = sc.nextLine().trim();
        tasks.add(new Task(nextId++, title, desc, prio, due, false));
        save();
        System.out.println("✓ Добавлена");
    }

    static void listAll() {
        if (tasks.isEmpty()) { System.out.println("Нет задач"); return; }
        for (Task t : tasks) {
            String status = t.completed ? "[✓]" : "[ ]";
            System.out.printf("%s %d: %s (%s) - %s [до %s]%n", status, t.id, t.title, t.priority, t.description, t.due);
        }
    }

    static void toggleComplete(Scanner sc) throws IOException {
        System.out.print("ID: ");
        int id = Integer.parseInt(sc.nextLine());
        for (Task t : tasks) {
            if (t.id == id) {
                t.completed = !t.completed;
                save();
                System.out.println("Статус изменён");
                return;
            }
        }
        System.out.println("Не найдено");
    }

    static void deleteTask(Scanner sc) throws IOException {
        System.out.print("ID для удаления: ");
        int id = Integer.parseInt(sc.nextLine());
        boolean removed = tasks.removeIf(t -> t.id == id);
        if (removed) { save(); System.out.println("Удалено"); }
        else System.out.println("Не найдено");
    }

    static void filterMenu(Scanner sc) {
        System.out.println("1. Активные\n2. Выполненные");
        String choice = sc.nextLine();
        for (Task t : tasks) {
            if (choice.equals("1") && !t.completed) System.out.printf("%d: %s%n", t.id, t.title);
            if (choice.equals("2") && t.completed) System.out.printf("%d: %s%n", t.id, t.title);
        }
    }
}
